File: 0202-happy-number/0202-happy-number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        
        
        s=set()
        sum=0
        temp=n
        while(n):
            
            sum+=(n%10)*(n%10)
            n//=10
            if n==0 and sum==1:
                    return True
            
            if n==0 and sum!=1 and temp not in s:
                #print(sum)
                s.add(temp) 
                n=sum
                temp=n
                sum=0
                
            elif n==0 and sum!=1 and temp in s:
                    return False
                
                
        return False
            
            
            
            

File: 0202-happy-number/0202-happy-number.cpp
class Solution {
public:
    bool isHappy(int n) {
        
                unordered_map<int,int>gzip;
        
                if(n==1)return true;
            
                
        
                while(gzip.find(n)==gzip.end()){
                        gzip[n]=1;
                        int sum=0;
                        while(n){
                            
                                 int last_idx=n%10;
                                n/=10;
                                sum+=pow(last_idx,2);
                                
                            
                        }
                        
                        n=sum;   
                        cout<<sum<<endl;
                        
                        if(n==1) return true;
                    
                }
        
                return false;
        
    }
};